cmd/boot-aws/main.go

```diff
@@ -0,0 +1,503 @@
+		return
+	}
+	fmt.Printf("IDs for any newly created resources are stored in %s. Use the teardown command to clean them up.\n", resourcesFile)
+	if err = resfile.Close(); err != nil {
+		fmt.Fprintf(os.Stderr, "error closing resources file: %s\n", err.Error())
+		fnerr = err
+		return
+	}
+}
+
+func doTeardown(aws *awscloud.AWS, res *resources) error {
+	if res.InstanceID != nil {
+		fmt.Printf("terminating instance %s\n", *res.InstanceID)
+		if _, err := aws.TerminateInstanceEC2(res.InstanceID); err != nil {
+			return fmt.Errorf("failed to terminate instance: %v", err)
+		}
+	}
+
+	if res.SecurityGroup != nil {
+		fmt.Printf("deleting security group %s\n", *res.SecurityGroup)
+		if _, err := aws.DeleteSecurityGroupEC2(res.SecurityGroup); err != nil {
+			return fmt.Errorf("cannot delete the security group: %v", err)
+		}
+	}
+
+	if res.AMI != nil {
+		fmt.Printf("deleting EC2 image %s and snapshot %s\n", *res.AMI, *res.Snapshot)
+		if err := aws.DeleteEC2Image(res.AMI, res.Snapshot); err != nil {
+			return fmt.Errorf("failed to deregister image: %v", err)
+		}
+	}
+	return nil
+}
+
+func teardown(cmd *cobra.Command, args []string) {
+	var fnerr error
+	defer func() { exitCheck(fnerr) }()
+
+	flags := cmd.Flags()
+
+	a, err := newClientFromArgs(flags)
+	if err != nil {
+		fnerr = err
+		return
+	}
+
+	resourcesFile, err := flags.GetString("resourcefile")
+	if err != nil {
+		return
+	}
+
+	res := &resources{}
+	resfile, err := os.Open(resourcesFile)
+	if err != nil {
+		fnerr = fmt.Errorf("failed to open resources file: %s", err.Error())
+		return
+	}
+	resdata, err := io.ReadAll(resfile)
+	if err != nil {
+		fnerr = fmt.Errorf("failed to read resources file: %s", err.Error())
+		return
+	}
+	if err := json.Unmarshal(resdata, res); err != nil {
+		fnerr = fmt.Errorf("failed to unmarshal resources data: %s", err.Error())
+		return
+	}
+
+	fnerr = doTeardown(a, res)
+}
+
+func doRunExec(a *awscloud.AWS, filename string, flags *pflag.FlagSet, res *resources) error {
+	sshKey, err := flags.GetString("ssh-key")
+	if err != nil {
+		return err
+	}
+
+	username, err := flags.GetString("username")
+	if err != nil {
+		return err
+	}
+
+	tmpdir, err := os.MkdirTemp("", "boot-test-*")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tmpdir)
+
+	hostsfile := filepath.Join(tmpdir, "known_hosts")
+	ip, err := a.GetInstanceAddress(res.InstanceID)
+	if err != nil {
+		return err
+	}
+	if err := keyscan(ip, hostsfile); err != nil {
+		return err
+	}
+
+	// TODO: remove this assumption and add a private key flag
+	key := strings.TrimSuffix(sshKey, ".pub")
+
+	// ssh into the remote machine and exit immediately to check connection
+	if err := sshRun(ip, username, key, hostsfile, "exit"); err != nil {
+		return err
+	}
+
+	// copy the executable
+	if err := scpFile(ip, username, key, hostsfile, filename); err != nil {
+		return err
+	}
+
+	// run the executable
+	return sshRun(ip, username, key, hostsfile, fmt.Sprintf("./%s", filename))
```

Comment:
You need to use `filepath.Base(filename)` here as well. It may be IMO cleaner to extract the base name here and pass the same value as `destination` to the `scpFile()` and `ssRun()`, instead of extracting the filename in `scpFile()`. My reasoning is that you should be able to expect that the provided destination passed to `scpFile()` will be used as is, not modified.

Reply:
Yeah, good catch.  Late night hacking fail :/

Reply:
Done!